### ISSUE TITLE
Deprecate DSL syntax in identity

### ISSUE THREAD

**Issue by theofidry** [nelmio/alice#512]
I can see from a test that we support:

```yaml
User:
  user1:
    'username': 'testuser'
  user2:
    username: '<(@user->username')>
  user3:
    username: '<(@user1->username . "_" . @user2->username')>
```

The issue I'm having with it is how much do we support in those expressions? It is quite easy for the identity function to have access to the *scope*:

```yaml
Dummy:
  dummy:
    'foo': 'bar'
    'foo_val': '<($foo)>'
```

As this is plain PHP. However the first example is a different matter. Because we need to parses strings, we can't really accept *anything*, supporting all the DSL features + bits of PHP would be a complete nightmare. We can reasonably support *some* features (like access to the variables), maybe the fixture references, but I don't think we should for the others. They are edge cases that should be solved via a custom Faker provider IMO.

@tshelburne what's your opinion on that matter?

**Comment by tshelburne** [comment]
Could you just parse the statement and turn it into PHP - effectively: 

```
<(@user1->username . "_" . @user2->username')>

// becomes
<($instances['user1']->username . "_" . $instances['user2']->username)>
```

**Comment by theofidry** [comment]
I had an idea along those lines if we wanted to support the fixture references in the identity expressions. But:

1. I would call it a "compromise", because as for access to the variables, this is something very useful but that doesn't bring much complexity and relatively easy to handle (little risk of syntax conflicts, it feels relatively natural)
2. If we were to do that there would still be a difference:

```yaml
Dummy:
  dummy:
    foo: 'bar'
  dummy_with_identity:
    foo: '<(@dummy->foo)>'
  dummy_without_identity:
    foo: '@dummy->foo'
```

In this example, `@dummy->foo` would mean get the `foo` property of `@dummy`, i.e. potentially via a getter. In the case of identity, if `foo` is private this would fail. This could still be ok as you would be able to use PHP to solve that:

```yaml
Dummy:
  dummy:
    foo: 'bar'
  dummy_with_identity:
    foo: '<(@dummy->getFoo())>'
  dummy_without_identity:
    foo: '@dummy->foo'
```

However this is a subtle difference that can confuse the user.

So from your comment I guess you would be for supporting the references and I would be ok with it. But my other points still stands: where do we draw the line? I would be for supporting only variables and references, nothing else.

**Comment by tshelburne** [comment]
That makes sense to me - I agree with your previous statement, anything more should be its own provider. Then the onus is on us to ensure provider registration is clean and simple.

**Comment by theofidry** [comment]
Closed via #560
